### ISSUE TITLE
ETA for OpenAPI 3 support?

### ISSUE THREAD

**Issue by moon0326** [BigstickCarpet/swagger-parser#72]
Hi!

Thank you for such a good package :) I've been using it for my projects.

Is there an ETA for OpenAPI 3 support? I know that it's coming, but just wanted to know if there is an ETA.

Thanks!
Moon

**Comment by BigstickCarpet** [comment]
Hi.  Sorry for not having a solid ETA date.  I had originally hoped to have it done by now, but then work got really hectic all of a sudden, so I haven't had any free time lately.  I hope to resume work on Swagger Parser next week.

**Comment by moon0326** [comment]
@BigstickCarpet thank you for the reply :) Have a good weekend!

**Comment by lcofre** [comment]
Great @BigstickCarpet !

Thanks very much for your work. I am in the process of migrating to 3.0.0 so your project update is invaluable!

**Comment by hugoduraes** [comment]
Can you make a beta release?

**Comment by AndyBarron** [comment]
any updates on this? anything specific that the community could throw together a PR for?

**Comment by BigstickCarpet** [comment]
Hi.  Please see [my comment here](https://github.com/BigstickCarpet/swagger-express-middleware/issues/81#issuecomment-341596208).  I was unable to get any work done for a couple months because of work and family commitments, but I'm back on it now.

**Comment by AndyBarron** [comment]
All good, take your time! We all appreciate your work on this. Just let us know if we can help with anything specific :)

**Comment by gonenduk** [comment]
Can't wait for an update :)

**Comment by sandorfr** [comment]
Is there any way we can help move this forward so we could have an alpha soon?

**Comment by sandorfr** [comment]
I don’t know about others, but I grow impatient on this one. I have the feeling the only way to move this forward would be through a PR but before doing duplicated work it would be great to know if it is started somewhere?

**Comment by gonenduk** [comment]
Any updates? ETA? Some news?

**Comment by gkoberger** [comment]
Is there any updates on this, or alternative packages we should be looking at?

We're happy to help fund development of this, since we rely on the package so much. Is that something that could help speed things up? We've looked into contributing ourselves, however we didn't want to duplicate work and wasn't sure how to make changes across multiple modules.

If there's a way we can help fund this, let me know! greg at readme dot io :)

(And thanks for building this... we love the package 💙)

**Comment by SomeoneRandom42** [comment]
Yeah I am with gkoberger. We really enjoy having the docs in line with the endpoints in comments and would love to keep using this package, but the new version of Open API adds many features we desperately want to upgrade.

If there is anything we can do please let me know.

**Comment by rssteffey** [comment]
I would also be up for helping expand to 3.0 support.  The spec allows for a lot more flexibility in our APIs here and we're pretty dependent on this repo

**Comment by gonenduk** [comment]
@gkoberger - maybe you should offer to fund it and some motivated people could take the project and work on it.

**Comment by ztheory** [comment]
@gonenduk - Pretty sure that's exactly what he did.

+1 for this.

**Comment by gkoberger** [comment]
I reached out to @BigstickCarpet and offered to finance it, however unfortunately he wasn't able to commit to anything due to work commitments. I completely understand, but we really needed OAS3 support!

So, we created `oas-normalize`, which accepts both Swagger 2 and OAS 3 (YAML or JSON, via string or URL or blob or whatever), and validates them. It has a similar API to `swagger-parser`, and uses `swagger-parser` under the hood for Swagger 2. (It uses `swagger2openapi` for validating OAS 3). It also uses @BigstickCarpet's `json-schema-ref-parser` for all the resolving of `$refs`.

https://github.com/readmeio/oas-normalize

Here's how to use it:

```
const OAS = require('oas-normalize');

// Use a URL, string, pathname, JSON blob, whatever
const oas = new OAS('https://raw.githubusercontent.com/OAI/OpenAPI-Specification/master/examples/v3.0/petstore-expanded.yaml');

oas.validate((err, spec) => {
  if (err) {
    console.log(err.errors);
    return;
  }
  console.log(spec); // spec will always be JSON, and valid
});
```

Currently it's backend-only, but we're going to try to fix that.

Let us know if you run into issues!

**Comment by gonenduk** [comment]
Great job!
Next step would be to use it in swagger-express-middleware.
If the interface is the same then it will be easy.

**Comment by MikeRalphson** [comment]
@gkoberger happy to help if `swagger2openapi ` needs any work done (it being my baby).

**Comment by BigstickCarpet** [comment]
I just [released Swagger Parser v5.0.0](https://github.com/BigstickCarpet/swagger-parser/blob/master/CHANGELOG.md), which includes initial support for OpenAPI 3.0. 👏🎉🎆

A big "Thank You!" to [Leo Long](https://github.com/yujunlong2000) for doing the work and submitting [PR #88](https://github.com/BigstickCarpet/swagger-parser/pull/88). 👍 🥇 💯
